### ISSUE TITLE
Multiple jsonb indexes fails

### ISSUE THREAD

**Issue by tgburrin** [pgpartman/pg_partman#232]
I'm working with a pg 10.4 with native partitioning, pg_partman 3.2.1, and the following example setup

`create table tgburrin.jsonb_table (id text not null, created_on timestamptz not null, object jsonb not null) partition by range (created_on);`

`SELECT partman.create_parent('tgburrin.jsonb_table','created_on','native','monthly');`

`create index on partman.template_tgburrin_jsonb_table ((object #>> '{field1}'));`
`create index on partman.template_tgburrin_jsonb_table ((object #>> '{field2}'));`

reapply_indexes.py fails with the following error(s):

```Jobs left in queue: 1
CREATE INDEX jsonb_table_p2019_02_idx ON "tgburrin"."jsonb_table_p2019_02" USING btree (((object #>> '{field2}'::text[])))
CREATE INDEX jsonb_table_p2019_02_idx ON "tgburrin"."jsonb_table_p2019_02" USING btree (((object #>> '{field1}'::text[])))
Process Process-9:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 267, in _bootstrap
    self.run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/postgresql/bin/reapply_indexes.py", line 257, in reindex_proc
    create_index(conn, partman_schema, child_schemaname, child_tablename, child_index_list, parent, parent_index_list)
  File "/usr/local/postgresql/bin/reapply_indexes.py", line 100, in create_index
    cur.execute(statement)
ProgrammingError: relation "jsonb_table_p2019_02_idx" already exists
```
It appears that reapply_indexes relies on specific field name(s) to create the name of the index on child tables.  Can this be updated to support functional indexes?

Would it be possible for reapply_indexes to split off the template/parent table name and use the suffix generated there as the suffix for all the partitions rather than attempting to calculate it from the fields within the index?

**Comment by keithf4** [comment]
Functional indexes should be supported. It parses out the index definition from running pg_get_indexdef(indexrelid) and takes everything after the USING clause as its definition. It should also be accounting for duplicate index names as well. It first checks to see if that index already exists and if it does, adds a counter to the name.

I'll try and test things out and see what's going on.

**Comment by keithf4** [comment]
Just confirming that I was able to reproduce this. Not sure I'll have a fix for it in my next release coming out alongside the PG11 release candidate. But I'll hopefully have a fix soon after if not.

**Comment by tgburrin** [comment]
Decided to try my hand at this.  What are your thoughts on the following as being a potential solution?
It seems like it should cover a majority of use cases.  The only extreme edge case I could think of is where a user specified index name could collide with a automatically determined one.

```63d62
<             # Index field names could not be extracted
67,78c66
<                 index_name += "_idx"
< 
<             else:
<                 # For default naming, we can attempt to replace the parent table with the child
<                 index_name = i[3].replace(parent_tablename, child_tablename)
< 
<                 # An index name may have been explicitly named on the template table
<                 # attempt to concatenate the two
<                 if index_name == i[3]:
<                     index_name = "_".join([child_tablename, i[3]])
< 
<             sql = "SELECT " + partman_schema + ".check_name_length('" + index_name + "')"
---
>             sql = "SELECT " + partman_schema + ".check_name_length('" + index_name + "', p_suffix := '_idx')"
107d94
````

**Comment by keithf4** [comment]
Still haven't gotten a chance to really dig into it. I don't think your fix completely solves the issue. Problem is for some reason the catalog lookup to see if that index name already exists isn't working so it will just add an incremental counter on to the next one.

However, if you're able to upgrade to PG11, this issue has mostly been solved since most indexes are now automatically inherited from the parent table in native partitioning. pg_partman 4.0.0 has been updated to handle this. The only thing the template table is now needed for is primary keys and/or unique keys that don't include the partition column. Right now my reapply_index.py script doesn't take that into account and I'm not quite sure how I'm going to update it to handle things.

**Comment by keithf4** [comment]
Figures as soon as I reply to this, I find the problem. Issue is that I had the wrong schema being fed as a parameter when checking for the child index existence. This only became an issue with the template tables because it can exist in a different schema than the partition table itself. Below is a diff with a fix. It's just a single parameter change to an execute call.
```
diff --git a/bin/common/reapply_indexes.py b/bin/common/reapply_indexes.py
index e6f97f1..9d1cab0 100755
--- a/bin/common/reapply_indexes.py
+++ b/bin/common/reapply_indexes.py
@@ -71,7 +71,7 @@ def create_index(conn, partman_schema, child_schemaname, child_tablename, child_
             while True:
                 sql = "SELECT count(*) FROM pg_class c JOIN pg_namespace n ON c.relnamespace = n.oid WHERE n.nspname = %s AND c.relname = %s"
                 cur = conn.cursor()
-                cur.execute(sql, [parent_schemaname, index_name])
+                cur.execute(sql, [child_schemaname, index_name])
                 index_exists = cur.fetchone()[0]
                 if index_exists != None and index_exists > 0:
                     index_name = child_tablename
```
Just note that if you do upgrade to 11, the reapply index script will not work properly if you're using native index inheritance. Since it's only looking at the template table, it will drop anything that doesn't exist there.

**Comment by keithf4** [comment]
This issue has been fixed in version 4.1.0. Thank you for reporting it!
